### PR TITLE
reuse SAMRAIDataCache objects between FEDataManager instances

### PR DESCRIPTION
**Pull request by boyceg**
Allow callers to share one (or more) `SAMRAIDataCache` objects between multiple `FEDataManager`s.

### CODE REVIEW

**Comment by boyceg** [comment]
Addresses #733.

**Comment by drwells** [comment]
We might want to just break the signature of that function: I really doubt its being called in user codes.

I want to wrap up some work on #654 before reviewing this carefully.

File: ibtk/include/ibtk/FEDataManager.h [hunk 1]
```diff
@@ -454,6 +456,7 @@ class FEDataManager : public SAMRAI::tbox::Serializable, public SAMRAI::mesh::St
                const InterpSpec& default_interp_spec,
                const SpreadSpec& default_spread_spec,
                const SAMRAI::hier::IntVector<NDIM>& min_ghost_width = SAMRAI::hier::IntVector<NDIM>(0),
+               std::shared_ptr<SAMRAIDataCache> eulerian_data_cache = nullptr,
                bool register_for_restart = true);
```
Comment (drwells) [inline ibtk/include/ibtk/FEDataManager.h]: This is the backwards compatible overload; I think that we should keep this one as-is and add the argument to the two newer ones instead.

Comment (drwells) [reply ibtk/include/ibtk/FEDataManager.h]: The two other overloads are the ones that IBFEMethod uses anyway, not this one.

File: tests/IBFE/explicit_ex4.cpp [hunk 2]
```diff
@@ -366,11 +369,15 @@ main(int argc, char** argv)
...
+            IntVector<NDIM> min_ghost_width = 0;
+            std::shared_ptr<IBTK::SAMRAIDataCache> eulerian_data_cache(new IBTK::SAMRAIDataCache());
             other_manager = FEDataManager::getManager(fe_data_manager->getFEData(),
                                                       "cloned_fe_data_manager",
                                                       fe_data_manager->getDefaultInterpSpec(),
                                                       fe_data_manager->getDefaultSpreadSpec(),
-                                                      spec);
+                                                      spec,
+                                                      min_ghost_width,
+                                                      eulerian_data_cache);
```
Comment (drwells) [inline tests/IBFE/explicit_ex4.cpp]: Is modifying this test necessary? Would it work in its current form?

File: src/IB/IBFEMethod.cpp [hunk 3]
```diff
@@ -1143,19 +1143,28 @@ IBFEMethod::initializeFEEquationSystems()
...
         {
+            if (!d_scratch_eulerian_data_cache) d_scratch_eulerian_data_cache.reset(new SAMRAIDataCache());
             d_scratch_fe_data_managers[part] = FEDataManager::getManager(d_primary_fe_data_managers[part]->getFEData(),
                                                                          manager_name + "::scratch",
                                                                          d_interp_spec[part],
                                                                          d_spread_spec[part],
-                                                                         d_workload_spec[part]);
+                                                                         d_workload_spec[part],
+                                                                         min_ghost_width,
+                                                                         d_scratch_eulerian_data_cache);
```
Comment (drwells) [inline src/IB/IBFEMethod.cpp]: :heavy_check_mark:

File: include/ibamr/IBFEMethod.h [hunk 4]
```diff
@@ -781,6 +782,11 @@ class IBFEMethod : public IBStrategy
     SAMRAI::tbox::Pointer<SAMRAI::mesh::GriddingAlgorithm<NDIM> > d_gridding_alg;
     bool d_is_initialized = false;
 
+    /*
+     * Scratch data caching objects.
```
Comment (drwells) [inline include/ibamr/IBFEMethod.h]: Could you add some more comments explaining that these are used by different sets of `FEDataManager`s?

File: tests/IBFE/explicit_ex4.cpp (hunk 2 above)
Comment (boyceg) [reply tests/IBFE/explicit_ex4.cpp]: Nope, not necessary. I will revert.

Comment (boyceg) [reply tests/IBFE/explicit_ex4.cpp]: It is not a necessary change. I will revert.

File: include/ibamr/IBFEMethod.h (hunk 4 above)
Comment (boyceg) [reply include/ibamr/IBFEMethod.h]: Sure.

**Pull request merged**